File: src/upros_odometry/src/ros_odom.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <nav_msgs/Odometry.h>
#include <nav_msgs/Path.h>

class OdomTracker
{
public:
    OdomTracker()
    {
        // 初始化ROS节点
        ros::NodeHandle nh;
        // 初始化订阅和发布的话题
        odom_sub_ = nh.subscribe("/odom", 10, &OdomTracker::odomCallback, this);
        path_pub_ = nh.advertise<nav_msgs::Path>("/odom_path", 10);
        // 设置默认的机器人位姿
        current_pose_.pose.pose.position.x = 0.0;
        current_pose_.pose.pose.position.y = 0.0;
        current_pose_.pose.pose.orientation.w = 1.0;
    }

private:
    ros::Subscriber odom_sub_;
    ros::Subscriber cmd_vel_sub_;
    ros::Publisher path_pub_;
    nav_msgs::Odometry current_pose_;
    nav_msgs::Path odom_path_;

    void odomCallback(const nav_msgs::Odometry::ConstPtr &msg)
    {
        // 更新机器人的位姿
        current_pose_ = *msg;
        // 将机器人的位姿添加到轨迹中
        odom_path_.header = current_pose_.header;
        geometry_msgs::PoseStamped pose;
        pose.pose.position.x = current_pose_.pose.pose.position.x;
        pose.pose.position.y = current_pose_.pose.pose.position.y;
        odom_path_.poses.push_back(pose);
        // 发布轨迹信息
        path_pub_.publish(odom_path_);
    }
};

int main(int argc, char **argv)
{
    // 初始化ROS节点
    ros::init(argc, argv, "odom_tracker");
    // 创建OdomTracker对象
    OdomTracker tracker;
    // 进入ROS循环
    ros::spin();
    return 0;
}

File: src/upros_odometry/scripts/ros_odom.py
#!/usr/bin/env python3

import rospy
from geometry_msgs.msg import Pose, Quaternion, Point, Twist
from nav_msgs.msg import Odometry, Path

class OdomTracker:
    def __init__(self):
        # 初始化ROS节点和发布者
        rospy.init_node('odom_tracker')
        self.path_pub = rospy.Publisher('/odom_path', Path, queue_size=10)
        # 设置默认的机器人位姿
        self.current_pose = Odometry()
        self.current_pose.pose.pose.position.x = 0.0
        self.current_pose.pose.pose.position.y = 0.0
        self.current_pose.pose.pose.orientation.w = 1.0

    def odom_callback(self, msg):
        # 更新机器人的位姿
        self.current_pose = msg
        # 将机器人的位姿添加到轨迹中
        path = Path()
        path.header = self.current_pose.header
        path.poses.append(self.current_pose.pose.pose)
        # 发布轨迹信息
        self.path_pub.publish(path)

if __name__ == '__main__':
    try:
        # 创建OdomTracker对象
        tracker = OdomTracker()
        # 订阅/odom话题，并设置回调函数为odom_callback
        rospy.Subscriber('/odom', Odometry, tracker.odom_callback)
        # 进入ROS循环
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
